src/draggable-list/index.tsx

```diff
@@ -0,0 +1,177 @@
+
+	const buildKey = (key: string) => `${keyRoot}-${key}`;
+
+	const renderItem = (item: DraggableListChild, index: number) => {
+		const itemDrag = drag.get(buildKey(item.key));
+		let currentPosition;
+		let padTop = false;
+		let padBottom = false;
+
+		// If this is a new drag, cache the starting item position, index, and height
+		if (itemDrag.start) {
+			const x = itemDrag.start.client.x - itemDrag.start.offset.x;
+			const y = itemDrag.start.client.y - itemDrag.start.offset.y;
+			startPosition = icache.set('startPosition', { x, y });
+			currentPosition = { x, y };
+			dragIndex = icache.set('dragIndex', index);
+			itemHeight = icache.set('itemHeight', dimensions.get(buildKey(item.key)).offset.height);
+			itemWidth = icache.set('itemWidth', dimensions.get(buildKey(item.key)).size.width);
+		}
+		// If this is an existing drag, use the last known position
+		else if (itemDrag.isDragging) {
+			currentPosition = lastPosition;
+		}
+
+		// Add top padding to the item immediately following the item being dragged
+		if (dragIndex !== undefined) {
+			if (
+				(renderIndex === undefined && index === dragIndex + 1) ||
+				(renderIndex !== undefined && renderIndex < dragIndex && index === renderIndex) ||
+				(renderIndex !== undefined && renderIndex >= dragIndex && index === renderIndex + 1)
+			) {
+				padTop = true;
+			}
+		}
+
+		// Add bottom padding to the end of the list if necessary
+		if (
+			renderIndex !== undefined &&
+			renderIndex === items.length - 1 &&
+			index === renderIndex
+		) {
+			padBottom = true;
+		}
+
+		// If this item is not being dragged, return a static list item
+		if (!itemDrag.isDragging || !startPosition || !currentPosition) {
+			return (
+				<li
+					classes={[
+						themeCss.item,
+						padTop && themeCss.paddedItemTop,
+						padBottom && themeCss.paddedItemBottom
+					]}
+					key={buildKey(item.key)}
+					styles={{
+						marginTop: padTop ? `${itemHeight}px` : undefined,
+						marginBottom: padBottom ? `${itemHeight}px` : undefined
+					}}
+				>
+					{item.content}
+				</li>
+			);
+		}
+
+		// Update and cache the last known position and the visual index
+		currentPosition.x += itemDrag.delta.x;
+		currentPosition.y += itemDrag.delta.y;
+		icache.set('lastPosition', currentPosition);
+		const offset = itemHeight
+			? Math.round((startPosition.y - currentPosition.y) / -itemHeight)
+			: 0;
+		icache.set('visualIndex', index + offset);
+
+		return (
+			<li
+				classes={[themeCss.item, themeCss.activeItem]}
+				key={buildKey(item.key)}
+				styles={{
+					left: lockAxis ? undefined : `${currentPosition.x}px`,
+					position: 'fixed',
+					top: `${currentPosition.y}px`,
+					width: `${itemWidth}px`,
+					zIndex: '100'
+				}}
+			>
+				{item.content}
+			</li>
+		);
+	};
+
+	const handleDragEnd = () => {
+		if (dragIndex === undefined || renderIndex === undefined) {
+			return;
+		}
+		onMove(dragIndex || 0, renderIndex || 0);
+		icache.clear();
+		icache.set('keyRoot', uuid());
```

Comment:
think i mentioned this elsewhere but if you must have a unique `id`, use the `id` from the factory, no need for uuid or to cache it.

Reply:
See comment above. When a drag starts and during a drag, the `keyRoot` must remain the same. Only once the drag ends do we change the `keyRoot` to selectively re-render any downstream node that relies on it. AFAIK, `id` does not provide the ability to reliably and explicitly generate a new, unique ID on-demand. It only guarantees no collisions with other widgets.